### ISSUE TITLE
ML features: baked SAT residual + YAML recipes (Approach B+C)

### ISSUE THREAD

**Issue by bbartling** [bbartling/open-fdd#104]
# ML features: Approach B + C (baked SAT residual + YAML recipes)

Epic / design issue: add **machine learning** capabilities aligned with `examples/AHU/RTU7_machine_learning.ipynb` **without arbitrary user-uploaded Python**. Two tracks:

- **Approach B — Baked modules:** versioned, reviewed Python in `open_fdd` (e.g. SAT residual regressor with sklearn).
- **Approach C — YAML recipes:** declarative config (target, features, fan mask, quantile, CV, retrain policy) — same operational feel as fault rule YAML.

**Out of scope for this issue:** Approach A (user `.py` upload + sandbox worker) — separate epic if needed.

---

## 1. Goals

- Analysts combine **YAML rules** with **ML-style detection** on the **same timeseries context** the platform already uses (DB + equipment + lookback), not only in a notebook with a local CSV.
- Outputs use existing **`fault_results`** / **`fault_definitions`** so Grafana, Faults UI, and downloads stay coherent.
- **PyPI `open-fdd`:** keep default install slim; add optional **`open-fdd[ml]`** (sklearn) for library + notebook users.
- **Full stack (Docker):** FDD loop + API images install the ML extra.

---

## 2. Strategic fork (context)

| Approach | Idea | This issue |
|----------|------|------------|
| **A** | User-uploaded `.py` | Not here — high security / sandbox cost |
| **B** | Baked sklearn modules in repo | **Yes** — `open_fdd/engine/ml/` |
| **C** | Declarative YAML recipes | **Yes** — `stack/ml_recipes/*.yaml` |

---

## 3. Baseline (current repo)

| Layer | Today |
|-------|--------|
| **Engine** (`open_fdd.engine`) | YAML → pandas → flag columns. No training. |
| **Notebook** (`RTU7_machine_learning.ipynb`) | CSV → `RuleRunner` → fan-on mask → sklearn `TimeSeriesSplit` + `GridSearchCV` → residual @ 95th percentile → plots. All local/trusted code. |
| **Platform** (`open_fdd/platform/loop.py`) | Loads `rules_dir` YAML every run, `RuleRunner.run`, `results_from_runner_output` → `_write_fault_results`. No ML. |
| **Frontend** | Faults page: YAML via `/rules`. No ML. |

---

## 4. End-to-end data flow (target)

1. **Ingest** (unchanged): scrapers → `timeseries_readings`.
2. **Scope:** FDD loop selects site/equipment, lookback window (same as today).
3. **Build `df`:** `load_timeseries_for_equipment` / `load_timeseries_for_site` (unchanged).
4. **Rule stage:** `RuleRunner.run` → rule flags (unchanged).
5. **ML stage (new):** For each applicable recipe: train-if-due OR score-with-artifact → boolean fault series.
6. **Persist:** `FDDResult` rows with stable `fault_id` (e.g. `ml_<recipe_name>_flag`), **`evidence` jsonb** for threshold, MAE, R², `model_version`, `trained_at`.
7. **Sync:** upsert `fault_definitions` for ML faults (`category: ml`).
8. **Events (optional):** extend realtime topics if UI needs instant refresh.

**Training vs scoring:** Document clearly — **retrain** on schedule / thresholds; **score** every FDD loop (cheap: load joblib, predict).

---

### 5.1 Engine / shared logic (notebook + platform)

| Path | Purpose |
|------|---------|
| `open_fdd/engine/ml/__init__.py` | Public exports |
| `open_fdd/engine/ml/sat_residual.py` | Fan-on mask, dropna, `TimeSeriesSplit` + `GridSearchCV` (or fixed estimator), fit, predict, residual, quantile threshold → boolean series aligned to `df` |
| `open_fdd/engine/ml/recipe_schema.py` | Pydantic (or dataclasses) for one recipe file |
| `open_fdd/engine/ml/artifacts.py` | Save/load **joblib** + **metadata.json** (trained_at, sklearn version, features, MAE, R², threshold) |
| `open_fdd/tests/engine/ml/test_sat_residual.py` | Synthetic DataFrame tests |

**PyPI:** `pyproject.toml` optional-dependencies `ml = ["scikit-learn>=1.3", ...]`

### 5.2 Recipe files on disk (Approach C)

| Path | Purpose |
|------|---------|
| `stack/ml_recipes/*.yaml` | One file per recipe |
| `examples/AHU/ml_recipes/*.example.yaml` | Documented templates matching notebook |

**Illustrative recipe shape:**

```yaml
name: ml_sat_residual_rtu7
type: ml_sat_residual
version: 1
enabled: true
equipment_type: [Air_Handling_Unit]
target: Supply_Air_Temperature_Sensor
features:
  - Outside_Air_Temperature_Sensor
  - Return_Air_Temperature_Sensor
  - Mixed_Air_Temperature_Sensor
  - Supply_Fan_Speed_Command
fan:
  column: Supply_Fan_Status
  mode: threshold
  output_threshold: 10
training:
  lookback_days: 60
  min_rows: 500
  max_rows: 50000
  cv_splits: 3
  estimator: random_forest
scoring:
  residual_quantile: 0.95
retrain:
  interval_hours: 168
  min_new_samples_since_train: 10000
artifacts:
  subdir: rtu7_sat_residual
```

### 5.3 Platform loop (wiring)

| File | Changes |
|------|---------|
| `open_fdd/platform/config.py` | `ml_recipes_dir` (default `stack/ml_recipes`), `ml_artifacts_dir` (default `stack/ml_artifacts`), `ml_enabled` |
| `open_fdd/platform/loop.py` | After `runner.run`, for each site/equipment + applicable recipe: train-if-due / score; append `FDDResult` list; do not fail entire run on one recipe error |
| `open_fdd/platform/ml_sync.py` (new) | `load_ml_recipes`, `sync_ml_fault_definitions(recipes)` → `fault_definitions` |
| `open_fdd/platform/drivers/run_rule_loop.py` | Unchanged if it only calls `run_fdd_loop` |

### 5.4 Schema / DB

| Item | Action |
|------|--------|
| `fault_results` | Reuse columns; put ML context in **`evidence` jsonb** |
| `fault_definitions` | Upsert from recipes (`fault_id`, `name`, `category=ml`, …) |
| **Optional Phase 1.5:** `ml_model_registry` | `recipe_name`, `site_id`, `equipment_id`, `trained_at`, `artifact_path`, `metrics jsonb`, `active` |

### 5.5 API (Phase 1 optional; Phase 2 richer)

| Endpoint | Purpose |
|----------|---------|
| `GET /ml/recipes` | List recipes + summary |
| `GET /ml/recipes/{name}` | Raw YAML |
| `POST /ml/recipes` | Upload recipe (mirror `/rules`) — optional if git-only |
| `POST /ml/retrain` | Admin retrain trigger |
| `GET /ml/status` | Last train/score (needs registry or metadata scan) |

Phase 1 can ship **file-only recipes** + no new endpoints if needed.

### 5.6 Frontend

| Phase | Deliverable |
|-------|-------------|
| **MVP** | New faults appear via `fault_definitions`; Config **master switch** `OFDD_ML_ENABLED`; Faults page section **“ML recipes”** (list + link to docs) |
| **Phase 2** | Form for features (point picker), **Retrain now**, metrics (MAE, R², threshold) |
| **Phase 3** | Plots overlay rule vs ML flag |

---

## 6. Execution flow (per equipment, per FDD run)

1. Load timeseries → `df`.
2. `RuleRunner.run` → rule flags.
3. For each applicable **ML recipe**:
   - **Score (default every run):** If artifact exists and retrain not due → load joblib → predict on scoring window → residual vs stored **threshold** from training (prefer storing threshold at train time; match notebook semantics in code comments).
   - **Train (occasional):** If no artifact OR `now - last_train > interval_hours` OR new rows > `min_new_samples_since_train` → subset training window, fan mask, dropna, TimeSeriesSplit + grid, persist joblib + metadata.
4. Convert ML flags → `FDDResult` (same sparse True-only pattern as `results_from_runner_output` where possible).

---

## 7. How often models update (policy)

| Action | Default | Rationale |
|--------|---------|-----------|
| **Score** | Every FDD loop (~3 h) | Cheap |
| **Retrain** | Weekly (`interval_hours: 168`) | Balance drift vs CPU |
| **Extra trigger** | `min_new_samples_since_train` | After big backfill |
| **Manual** | `POST /ml/retrain` or touch-file | Commissioning |
| **Cap** | `max_rows` on training | Protect loop runtime |

**Docs must state:** training uses past data only; TimeSeriesSplit avoids leakage in CV.

---

## 8. `fault_id` naming

- Stable: `ml_<recipe_name>_flag` (must align with `fault_definitions` upsert).
- `sync_ml_fault_definitions` runs after loading recipes (parallel to `_sync_fault_definitions_from_rules`).

---

## 9. Docker / dependencies

- **fdd-loop** + **api** images: `pip install '.[ml]'` or equivalent.
- **Volumes:** persist `stack/ml_artifacts` (and optionally `ml_recipes` if UI upload).

---

## 10. Phased implementation checklist

**Sprint 1 — Library**
- [ ] `open_fdd/engine/ml/sat_residual.py` + tests
- [ ] `recipe_schema.py` + example YAML under `examples/AHU/ml_recipes/`
- [ ] Refactor `RTU7_machine_learning.ipynb` to call library (proves parity)

**Sprint 2 — Loop**
- [ ] `load_ml_recipes`, train/score, artifacts per `(site, equipment, recipe)`
- [ ] Emit `FDDResult` + `sync_ml_fault_definitions`
- [ ] `OFDD_ML_ENABLED` feature flag

**Sprint 3 — Ops & UI**
- [ ] Config page toggles + `docs/howto/ml_sat_residual.md`
- [ ] Faults page “ML recipes” section + API list if needed
- [ ] Optional `ml_model_registry` + `GET /ml/status`

**Sprint 4 — UX**
- [ ] Plots: overlay ML vs rule
- [ ] `POST /ml/retrain` + admin guard

---

## 11. Security note (for future reference)

User-uploaded Python (Approach A) requires a **sandboxed worker**, not `importlib` in API/FDD process. **This issue explicitly excludes A.**

---

## 12. Related files (existing)

- `examples/AHU/RTU7_machine_learning.ipynb`
- `open_fdd/platform/loop.py` — `run_fdd_loop`, `_write_fault_results`
- `open_fdd/schema/fdd_result.py` — `FDDResult`, `results_from_runner_output`
- `stack/sql/012_fault_definitions.sql`
- `frontend/src/components/pages/FaultsPage.tsx` — rules YAML UI pattern to mirror for ML section

---

*Issue body generated from architecture planning; implementation should split into sub-issues per sprint.*
